mindtrace/core/mindtrace/core/utils/hashing.py

```diff
@@ -0,0 +1,32 @@
+import hashlib
+from pathlib import Path
+
+
+def compute_dir_hash(directory_path: str | Path) -> str:
+    """Compute SHA256 hash of directory contents.
+
+    Hash is deterministic: files are sorted by path, then each file's content is hashed and combined. This ensures the
+    same directory always produces the same hash.
+
+    Args:
+        directory_path: Path to the directory to hash
+
+    Returns:
+        Hexadecimal SHA256 hash string
+    """
+    directory_path = Path(directory_path)
+    sha = hashlib.sha256()
+
+    # Sort files for deterministic hashing
+    # Only hash regular files, ignore directories
+    files = sorted([f for f in directory_path.rglob("*") if f.is_file()])
+
+    for file_path in files:
+        # Include relative path in hash for uniqueness
+        rel_path = file_path.relative_to(directory_path)
+        sha.update(str(rel_path).encode("utf-8"))
+
+        # Include file content
+        sha.update(file_path.read_bytes())
```

Comment:
we can stream in chunks to keep memory usage in check. some dry runs show it's faster as well.
```python
In [10]: %timeit compute_dir_hash("/home/vikrant/Downloads")  # 4.6GB, 78 files
3.11 s ± 8.93 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [11]: %timeit compute_dir_hash_streamed("/home/vikrant/Downloads")
2.25 s ± 3.91 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [12]: %timeit compute_dir_hash("/home/vikrant/models")  # 39GB, 2 files
25 s ± 30.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [13]: %timeit compute_dir_hash_streamed("/home/vikrant/models")
18.7 s ± 16.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

```suggestion
        with open(file_path, "rb") as fp:
            while True:
                chunk = fp.read(10 * (2 ** 20))  # 10 MB chunks
                if not chunk:
                    break
                sha.update(chunk)
```